keps/sig-node/4033-group-driver-detection-over-cri/README.md

```diff
@@ -0,0 +1,914 @@
+of a different kind of cgroups hierarchy used in them. Having kubelet and the
+container runtime to use non-matching cgroup drivers can cause hard to
+understand failures in container creation or inconsistent resource allocation
+on the node. This – two independent configuration settings for the same thing –
+is a common cause for user errors. Instead of having a split brain situation,
+there should be a single source of truth for the cgroup driver.
+
+### Goals
+
+- make kubelet automatically use the same cgroup driver as the container
+  runtime, making it unnecessary to specify cgroup driver in kubelet
+  configuration
+- maintain backward compatibility to run kubelet on top of an older container
+  runtime that doesn’t support this new feature
+
+### Non-Goals
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster administrator, I would like to simplify my node configuration by
+configuring fields just once, and not needing to synchronize between multiple
+processes.
+
+#### Story 2
+
+As a novice Kubernetes user, I would like to easily be able to start pods with
+all runtimes, even if they have differing opinions on defaults.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Field adoption could be considered a risk, though the CRI implementations work
+closely with SIG Node and the feature will move along with CRI implementation
+adoption.
+
+## Design Details
+
+### CRI API
+
+Extend the CRI runtime API to inform the kubelet which cgroup driver should be
+used. A new RuntimeConfig rpc is added to query the information.
+
+```diff
+ // Runtime service defines the public APIs for remote container runtimes
+ service RuntimeService {
+ ...
++    // RuntimeConfig returns configuration information of the runtime.
++    rpc RuntimeConfig(RuntimeConfigRequest) returns (RuntimeConfigResponse) {}
+ }
+ 
++message RuntimeConfigRequestRequest {}
+ 
++message RuntimeConfigResponse {
++    // Configuration information for Linux-based runtimes. This field contains global
++    // runtime configuration options that are not specific to runtime handlers.
++    LinuxRuntimeConfiguration linux = 1;
++}
+ 
++message LinuxRuntimeConfiguration {
++    // Cgroup driver to use
++    CgroupDriver cgroup_driver = 1;
++}
+ 
++enum CgroupDriver {
++    CGGROUPFS = 0;
++    SYSTEMD = 1;
++}
+```
+
+### Kubelet
+
+Kubelet will be modified to support the new field.
+
+If available the cgroup driver information received from the container runtime
+will take precedence over cgroupDriver setting from the kubelet config (or
+`--cgroup-driver` command line flag). If the runtime does not provide
+information about the cgroup driver, then kubelet will fall back to using its
```

Comment:
this is clear if the response is unimplemented, but what happens if it times out? how long do we wait try?

Reply:
I'm fine with anything, just we should document it so we are sure it has been discussed

Reply:
Added the following:

```
If supported by the runtime, RuntimeConfig query is expected to
+succeed, an error (error response or timeout) is regarded as a failed
+initialization of the runtime service and kubelet will exit with an error
+message and an error code.
```

@aojea @haircommander  WDYT?

Reply:
that works with me